File: cpp/okf/read.cpp
// Bundle reading — mirrors py/okf/okf.py::read_bundle.
//
// Parity rules: hidden directories skipped; only *.md files not starting with
// '.'; concepts sorted by bundle-relative forward-slash path in byte order.

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <sstream>

#include "okf.hpp"

namespace fs = std::filesystem;

namespace okf {
namespace {

std::string slashify(const fs::path& p) {
    std::string s = p.generic_string();  // forward slashes
    // Strip the Windows extended-length prefix if canonicalization added it.
    if (s.rfind("//?/", 0) == 0) s = s.substr(4);
    return s;
}

void walk_md(const fs::path& dir, std::vector<fs::path>& out, std::vector<fs::path>& all) {
    for (const fs::directory_entry& entry : fs::directory_iterator(dir)) {
        std::string name = entry.path().filename().string();
        if (entry.is_directory()) {
            if (!name.empty() && name[0] != '.') walk_md(entry.path(), out, all);
        } else if (!name.empty() && name[0] != '.') {
            all.push_back(entry.path());
            if (name.size() > 3 && name.compare(name.size() - 3, 3, ".md") == 0) {
                out.push_back(entry.path());
            }
        }
    }
}

std::optional<std::string> scalar(const json& fm, const char* key) {
    if (!fm.is_object()) return std::nullopt;
    auto it = fm.find(key);
    if (it == fm.end() || it->is_null() || it->is_array() || it->is_object()) {
        return std::nullopt;
    }
    if (it->is_string()) return it->get<std::string>();
    return it->dump();
}

}  // namespace

Bundle read_bundle(const std::string& root, const std::string& source_kind) {
    fs::path root_path = fs::canonical(fs::path(root));
    std::string root_str = slashify(root_path);
    std::vector<fs::path> files, all_paths;
    walk_md(root_path, files, all_paths);

    Bundle b;
    for (const fs::path& f : all_paths) {
        b.files.insert(slashify(fs::relative(f, root_path)));
    }
    b.root = root_str;
    b.source_kind = source_kind;
    b.concepts.reserve(files.size());
    for (const fs::path& f : files) {
        std::string rel = slashify(fs::relative(f, root_path));
        std::ifstream in(f, std::ios::binary);
        if (!in) throw OkfError("cannot read " + f.string());
        std::ostringstream ss;
        ss << in.rdbuf();
        Parsed p = parse_text(ss.str());

        Concept c;
        c.path = rel;
        std::size_t slash = rel.rfind('/');
        std::string base = slash == std::string::npos ? rel : rel.substr(slash + 1);
        c.reserved = RESERVED.count(base) > 0;
        c.type = scalar(p.meta, "type");
        c.title = scalar(p.meta, "title");
        c.description = scalar(p.meta, "description");
        c.resource = scalar(p.meta, "resource");
        c.tags = p.meta.is_object() && p.meta.contains("tags") ? p.meta["tags"] : json(nullptr);
        c.timestamp = scalar(p.meta, "timestamp");
        // OKF v0.2: fall back to `generated: {by, at}` when the legacy
        // `timestamp` is absent (spec section 13).
        if (!c.timestamp && p.meta.is_object() && p.meta.contains("generated")) {
            c.timestamp = scalar(p.meta["generated"], "at");
        }
        c.frontmatter = p.meta;
        c.parse_error = p.parse_error;
        c.links_raw = extract_links(p.body);
        c.wikilinks_raw = extract_wikilinks(p.body);
        c.content_hash = content_hash(p.body);
        c.body = std::move(p.body);
        b.concepts.push_back(std::move(c));
    }
    std::sort(b.concepts.begin(), b.concepts.end(),
              [](const Concept& a, const Concept& c2) { return a.path < c2.path; });
    for (const Concept& c : b.concepts) b.known.insert(c.path);
    for (const Concept& c : b.concepts) {
        if (c.path == "index.md") {
            b.okf_version = scalar(c.frontmatter, "okf_version");
            break;
        }
    }
    b.bundle_id = content_hash(root_str);
    return b;
}

}  // namespace okf

File: rust/okf-ingest/src/read.rs
//! Bundle reading — mirrors `py/okf/okf.py::read_bundle`.
//!
//! Parity rules: hidden directories (`.git`, `.github`, …) are skipped; only
//! `*.md` files not starting with `.`; concepts sorted by bundle-relative
//! forward-slash path in byte order (== code-point order == the catalog's
//! `ORDER BY path`).

use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf, MAIN_SEPARATOR as MAIN_SEP};

use crate::links::{extract_links, extract_wikilinks};
use crate::model::{Bundle, Concept, OkfError, SourceKind, RESERVED};
use crate::parse::{content_hash, meta_get, parse_text, scalar_str, yaml_to_json};

/// Canonicalize to a forward-slash string, stripping the Windows `\\?\` prefix.
fn canon_str(root: &Path) -> std::io::Result<String> {
    let c = fs::canonicalize(root)?;
    let s = c.to_string_lossy().replace('\\', "/");
    Ok(s.strip_prefix("//?/").map(str::to_string).unwrap_or(s))
}

fn walk_md(dir: &Path, out: &mut Vec<PathBuf>, all: &mut Vec<PathBuf>) -> std::io::Result<()> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let name = entry.file_name().to_string_lossy().to_string();
        let path = entry.path();
        if path.is_dir() {
            if !name.starts_with('.') {
                walk_md(&path, out, all)?;
            }
        } else if !name.starts_with('.') {
            all.push(path.clone());
            if name.ends_with(".md") {
                out.push(path);
            }
        }
    }
    Ok(())
}

/// Read an OKF bundle directory into memory.
pub fn read_bundle(
    root: &Path,
    bundle_id: Option<&str>,
    source_kind: SourceKind,
) -> Result<Bundle, OkfError> {
    let root_str = canon_str(root)?;
    let root_path = PathBuf::from(&root_str);
    let mut files = Vec::new();
    let mut all_paths = Vec::new();
    walk_md(&root_path, &mut files, &mut all_paths)?;
    let mut all_files: BTreeSet<String> = BTreeSet::new();
    for f in &all_paths {
        if let Ok(r) = f.strip_prefix(&root_path) {
            all_files.insert(r.to_string_lossy().replace(MAIN_SEP, "/"));
        }
    }

    let mut concepts: Vec<Concept> = Vec::with_capacity(files.len());
    for f in &files {
        let rel = f
            .strip_prefix(&root_path)
            .map_err(|e| OkfError::msg(e.to_string()))?
            .to_string_lossy()
            .replace('\\', "/");
        let text = fs::read_to_string(f)?;
        let p = parse_text(&text);
        let basename = rel.rsplit('/').next().unwrap_or(&rel);
        let (kind, title, description, resource, tags, timestamp, frontmatter) = match &p.meta {
            Some(h) => (
                meta_get(h, "type").and_then(scalar_str),
                meta_get(h, "title").and_then(scalar_str),
                meta_get(h, "description").and_then(scalar_str),
                meta_get(h, "resource").and_then(scalar_str),
                meta_get(h, "tags").map(yaml_to_json),
                // OKF v0.2: fall back to `generated: {by, at}` when the
                // legacy `timestamp` is absent (spec section 13).
                meta_get(h, "timestamp").and_then(scalar_str).or_else(|| {
                    meta_get(h, "generated").and_then(|g| match g {
                        yaml_rust2::Yaml::Hash(gh) => meta_get(gh, "at").and_then(scalar_str),
                        _ => None,
                    })
                }),
                {
                    let mut m = serde_json::Map::new();
                    for (k, v) in h.iter() {
                        m.insert(scalar_str(k).unwrap_or_default(), yaml_to_json(v));
                    }
                    Some(m)
                },
            ),
            None => (None, None, None, None, None, None, None),
        };
        concepts.push(Concept {
            reserved: RESERVED.contains(&basename),
            path: rel,
            kind,
            title,
            description,
            resource,
            tags,
            timestamp,
            links_raw: extract_links(&p.body),
            wikilinks_raw: extract_wikilinks(&p.body),
            content_hash: content_hash(&p.body),
            body: p.body,
            frontmatter,
            parse_error: p.err.map(str::to_string),
        });
    }
    concepts.sort_by(|a, b| a.path.cmp(&b.path));

    let known: BTreeSet<String> = concepts.iter().map(|c| c.path.clone()).collect();
    let okf_version = concepts
        .iter()
        .find(|c| c.path == "index.md")
        .and_then(|c| c.frontmatter.as_ref())
        .and_then(|fm| fm.get("okf_version"))
        .and_then(json_scalar);
    let bundle_id = match bundle_id {
        Some(b) => b.to_string(),
        None => content_hash(&root_str),
    };
    Ok(Bundle {
        bundle_id,
        root: root_str,
        okf_version,
        source_kind,
        concepts,
        known,
        files: all_files,
    })
}

/// JSON value -> string, mirroring `_s` on the frontmatter map (`None` for
/// arrays/objects/null). Kept private to `okf_version` lookup; scalar fields
/// come straight from YAML via [`scalar_str`].
fn json_scalar(v: &serde_json::Value) -> Option<String> {
    match v {
        serde_json::Value::String(s) => Some(s.clone()),
        serde_json::Value::Number(n) => Some(n.to_string()),
        serde_json::Value::Bool(b) => Some(b.to_string()),
        _ => None,
    }
}
